### PR TITLE
VMI controller - move unit tests to fake client (VirtualMachineInstanceInterface)

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
Use fake client instead of mocking in order to simulate real behavior.

The PR is exposing and fixing few tests: 
- [vmi ctrl: fix pod annotation patch tests](https://github.com/kubevirt/kubevirt/pull/11751/commits/c4819ab6466eeac5026c52f82603e87f82b8c7d0)
- [fix(vmi ctrl): remove empty activepods and volumestatus](https://github.com/kubevirt/kubevirt/pull/11751/commits/699c126a54ddbc3b99c8d9fc1f2fd04950a65421)
- [vmi ctrl: fix vmi status hotplug disk tests](https://github.com/kubevirt/kubevirt/pull/11751/commits/9cf28da2c16cb0634e2220309de4bd195923dda9)
- [vmi ctrl: fix invtsc and hyperv tests](https://github.com/kubevirt/kubevirt/pull/11751/commits/f3f4bf50117d45d06c7ca0329fbeb0c91c33db0f)

This PR also improved some tests:
- [fix(vmi ctrl): improve some tests](https://github.com/kubevirt/kubevirt/pull/11751/commits/4aae60e650a07d2c78920a8dc505081f5b14297d)

Fixes #

### Why we need it and why it was done in this way
Lower the bar for unit tests, eliminate false positives. The fake is standard for testing.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by fossedihelm** [comment]
/unhold
https://github.com/kubevirt/kubevirt/pull/11685 was merged

File: pkg/virt-controller/watch/vmi_test.go [hunk 1]
```diff
@@ -155,71 +155,67 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
...
 		})
 	}
 
-	shouldExpectVirtualMachinePendingState := func(vmi *virtv1.VirtualMachineInstance) {
-		vmiInterface.EXPECT().Update(context.Background(), gomock.Any(), metav1.UpdateOptions{}).Do(func(ctx context.Context, arg interface{}, options metav1.UpdateOptions) {
-			Expect(arg.(*virtv1.VirtualMachineInstance).Status.Phase).To(Equal(virtv1.Pending))
-			Expect(arg.(*virtv1.VirtualMachineInstance).Status.Conditions).To(ContainElement(MatchFields(IgnoreExtras,
-				Fields{"Type": Equal(virtv1.VirtualMachineInstanceReady)})))
-		}).Return(vmi, nil)
+	expectVMIBeInPhase := func(vmi *virtv1.VirtualMachineInstance, phase virtv1.VirtualMachineInstancePhase) {
```
Comment (xpivarc) [inline pkg/virt-controller/watch/vmi_test.go:170]: Please use `namespace` & `name`

File: pkg/virt-controller/watch/vmi_test.go [hunk 2]
```diff
@@ -4028,6 +4027,15 @@ func unmarkReady(vmi *virtv1.VirtualMachineInstance) {
 	kvcontroller.NewVirtualMachineInstanceConditionManager().RemoveCondition(vmi, virtv1.VirtualMachineInstanceConditionType(k8sv1.PodReady))
 }
 
+func addVolumeStatuses(vmi *virtv1.VirtualMachineInstance, volumeStatuses ...virtv1.VolumeStatus) *virtv1.VirtualMachineInstance {
+	if vmi.Status.VolumeStatus == nil {
+		vmi.Status.VolumeStatus = make([]virtv1.VolumeStatus, 0)
+	}
```
Comment (xpivarc) [inline pkg/virt-controller/watch/vmi_test.go:4092]: Isn't append safe?

File: pkg/virt-controller/watch/vmi_test.go [hunk 3]
```diff
@@ -1004,6 +1004,8 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 			addVirtualMachine(vmi)
 
 			controller.Execute()
+			Expect(virtClientset.Actions()).To(HaveLen(1)) //1 is the vmi creation
```
Comment (xpivarc) [inline pkg/virt-controller/watch/vmi_test.go:1005]: We should be able to do even better and be able to check what action. Meaning that we can code in that only the vmi creation happened.

File: pkg/virt-controller/watch/vmi_test.go [hunk 4]
```diff
@@ -106,46 +107,24 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
...
-	shouldExpectMultiplePodDeletions := func(pod *k8sv1.Pod, deletionCount *int) {
-		// Expect pod deletion
-		kubeClient.Fake.PrependReactor("delete", "pods", func(action testing.Action) (handled bool, obj k8sruntime.Object, err error) {
-			update, ok := action.(testing.DeleteAction)
-			Expect(ok).To(BeTrue())
-			Expect(pod.Namespace).To(Equal(update.GetNamespace()))
-			*deletionCount++
-			return true, nil, nil
-		})
+	expectPodDoesNotExist := func(pod *k8sv1.Pod) {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-controller/watch/vmi_test.go]: Done

File: pkg/virt-controller/watch/vmi_test.go [hunk 5]
```diff
@@ -341,6 +341,24 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 		mockQueue.ExpectAdds(1)
 		vmiSource.Add(vmi)
 		mockQueue.Wait()
+		_, err := virtClientset.KubevirtV1().VirtualMachineInstances(vmi.Namespace).Create(context.Background(), vmi, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	addPod := func(pod *k8sv1.Pod) {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-controller/watch/vmi_test.go:324]: Removed the podFeeder and the dataVolumeFeeder too.
Thank you!

File: pkg/virt-controller/watch/vmi_test.go [hunk 6]
```diff
@@ -341,6 +341,24 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 		mockQueue.ExpectAdds(1)
 		vmiSource.Add(vmi)
 		mockQueue.Wait()
+		_, err := virtClientset.KubevirtV1().VirtualMachineInstances(vmi.Namespace).Create(context.Background(), vmi, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-controller/watch/vmi_test.go:321]: We need to be sure that we check all tests. I know I have seen some test not requiring it but 99% of the time you do want to do this.

File: pkg/virt-controller/watch/vmi_test.go [hunk 7]
```diff
@@ -427,12 +424,9 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
...
 					return ""
 				},
-				Equal("/bin/bash -c echo bound PVCs"))
+				Equal("/bin/bash -c echo bound PVCs"),
+			)
 			shouldExpectMatchingPodCreation(vmi.UID, IsPodWithoutVmPayload)
-			kubeClient.Fake.PrependReactor("get", "persistentvolumeclaims", func(action testing.Action) (handled bool, obj k8sruntime.Object, err error) {
-				Fail("here")
-				return true, dvPVC, nil
-			})
```
[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-controller/watch/vmi_test.go:424]: Makes sense

File: pkg/virt-controller/watch/vmi_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-controller/watch/vmi_test.go:321]: I've checked the tests and looks like some tests don't require the vmi to be stored in the fakeclient. 
**BUT**
There aren't any tests that require the vmi **not** being stored in the fakeclient.
So, from my point of view, we are good as is. It also represents the reality.

File: pkg/virt-controller/watch/vmi_test.go [hunk 8]
```diff
@@ -1156,29 +1084,25 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
...
 				Expect(mockQueue.GetRateLimitedEnqueueCount()).To(BeZero())
-				addVirtualMachine(vmi)
-
-				// make sure that during next iteration we do not add the same condition again
-				vmiInterface.EXPECT().Update(context.Background(), gomock.Any(), metav1.UpdateOptions{}).Do(func(ctx context.Context, vmi *virtv1.VirtualMachineInstance, options metav1.UpdateOptions) {
-					expectConditions(vmi)
-				}).Return(vmi, nil)
+				mockQueue.ExpectAdds(1)
+				vmiSource.Add(vmi)
+				mockQueue.Wait()
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/vmi_test.go:1097]: Why not
```suggestion
				addVirtualMachine(vmi)
```
?

**Review by jean-edouard — COMMENTED** [review]
/approve
/hold
Thank you for the massive improvement!
I didn't inspect every single line, but I didn't see any significant test/check getting removed, and the tests pass, so I don't see a reason not to approve. (just one minor question in-line)
Feel free to remove the hold once every reviewer is satisfied with the code.

**Pull request merged**